finetune/nn/target_blocks.py

```diff
@@ -65,9 +65,76 @@ def cps_model(M, cps_mask, hidden, config):
        #mlm_loss *= (1 - tf.reduce_sum(tf.nn.softmax(logits) * tf.one_hot(cps_mask, 2, dtype=tf.float32), -1)) ** 2.0
     return {"loss": mlm_loss, "logits": logits}
 
+def masked_language_model_(*, X, M, mlm_weights, mlm_ids, mlm_positions, targets, embed_weights, hidden, config, reuse=None, train=False):
+    with tf.variable_scope('model/masked-language-model'):
+        batch, seq, feats = shape_list(hidden)
+        flat_offsets = tf.reshape(
+            tf.range(0, batch, dtype=tf.int32) * seq, [-1, 1]
+        )
+
+        not_padding = tf.reshape(mlm_weights, [-1]) > 1e-9
+        flat_positions = tf.boolean_mask(tf.reshape(mlm_positions + flat_offsets, [-1]), not_padding) # take off the padding entirely
+        gathered_hidden = tf.gather(tf.reshape(hidden, [batch * seq, feats]), flat_positions)
+        # Commenting this out and replacing with simple positional loss
+        mlm_ids = tf.boolean_mask(tf.reshape(mlm_ids, [-1]), not_padding)
+        # Unsure if this is directly applicable--dimension should be (batch, seq, 1))
+        # as far as I understand it, the not_padding stuff is generated from mlm_weights
+        # which should represent the mask.
+        targets = tf.boolean_mask(tf.reshape(targets, [-1,4]), not_padding)
+        final_proj = dense_with_custom_init(
+            gathered_hidden,
+            # Projection down to a single value for regression
+            config.n_embed,
+            activation=act_fns[config.act_fn],
+            kernel_initializer=tf.random_normal_initializer(stddev=config.weight_stddev),
+            name='dense',
+            custom=config.use_auxiliary_info and not (config.mlm_baseline or config.pos_injection),
+            pos_embed=config.n_context_embed_per_channel * config.context_dim,
+            proj_type='downward',
+            transpose_b=True
+        )
+        normed_proj = norm(final_proj, 'LayerNorm')
+        n_vocab = shape_list(embed_weights)[0]
+        output_bias = tf.get_variable(
+            #"output_bias",
+            "output_bias_posreg",
+            #shape=[n_vocab],
+            # We are regressing into 4D space...
+            # might be a good idea to do top left + offset?
+            shape=[4],
+            initializer=tf.zeros_initializer()
+        )
+        
+        # Going to add a new variable to project downwards
+        regression_matrix = tf.get_variable(
```

Comment:
I feel like rather than projecting out to n_vocab and projecting down you should just directly project down to the shape of the position outputs.

Reply:
Yeah, I suppose it isn't like I'm really preserving anything in the logits, so that seems like a change to try. My thought was that I wanted to preserve as much as the structure as possible but now that I think about it I doubt it's that relevant.

Reply:
There are decent reasons why you don't want this -- mainly the final layers of model are the ones that are most task specific that don't transfer as well. This is why you lop off the MLM head of bert at finetuning time -- those params aren't well suited to helping solve a classification task, for instance.